### PR TITLE
`Lifetime.observeEnded` and `take(during:)` fix.

### PR DESCRIPTION
**Pull request by andersio**
1. New methods and operator: ~~`attach`,~~ `observeEnded` ~~and `+=`~~.

   `ended` provides a full `Signal` interface that is overblown for the general use cases of `Lifetime`.

1. [Bugfix?] Safe observation ~~& deprecating `Lifetime.ended`.~~

   `ended.observeCompleted` does not handle the observations to a terminated `ended` Signal, and the action is simply ignored.

    [It should have been invoked regardless](https://github.com/ReactiveCocoa/ReactiveSwift/pull/229/files#diff-66075186b734efda335ddf4d9c209017R62), like how adding a `Disposable` to disposed `CompositeDisposable` would dispose of the disposable being added.

   ~~Composition of `Lifetime` is suggested to be done through lifetime operators instead, e.g. `and` and `or`, should the demand arise.~~

### CODE REVIEW

**Comment by sharplet** [comment]
I think I'm 👍  on `attach` and `observeEnded`, although I think `attach` could use a more descriptive name. But I think adding `+=` would be a step backwards in terms of clarity.

File: Tests/ReactiveSwiftTests/LifetimeSpec.swift [hunk 1]
```diff
@@ -1,34 +1,97 @@
...
-import ReactiveSwift
+@testable import ReactiveSwift
 import Result
 
 final class LifetimeSpec: QuickSpec {
 	override func spec() {
 		describe("Lifetime") {
-			it("should complete its lifetime ended signal when the it deinitializes") {
-				let object = MutableReference(TestObject())
+			describe("ended") {
```
Comment (sharplet) [inline Tests/ReactiveSwiftTests/LifetimeSpec.swift]: It's a nitpick, but do you think we can make this change without reindenting the file? It makes the diff a little hard to follow, and there's no shared setup/teardown code.

**Comment by mdiep** [comment]
I really like the simplicity of exposing `ended: Signal<(), NoError>`. Clear, unambiguous, and does very little to expand the API. I also think we should prefer `Signal`s to `Disposable`s.

If we're concerned about the safety issue you mentioned, adding the terminated events to `Observer` seems like a fine approach.

**Comment by inamiy** [comment]
I'm +1 for not exposing `ended: Signal<(), NoError>`.
It's hard to grasp from the type that only `ended.observeCompleted { ... }` will work.

But I don't get why disposable is needed as an argument of `func attach`.
I think just plain `() -> ()` is sufficient.

Also, I don't think `Observer(terminated:)` is necessary because `end` only does `sendCompleted`.
(Is there any cases of `sendInterrupted` or others?)

P.S. I think `Signal<Never, NoError>` describes a bit better than `Signal<(), NoError>`.

**Comment by andersio** [comment]
@inamiy When the lifetime has already terminated.

**Comment by andersio** [comment]
@mdiep #143 might have a stake in this, if it is still on the table.

**Comment by inamiy** [comment]
@andersio 
Ah, you are right. Lifetime may `sendInterrupted`.
And with #143, I got the idea of `func attach`.

**Comment by mdiep** [comment]
I still prefer `ended: Signal<(), NoError>`.

I can be 👍 to `observeEnded` if people find `ended` confusing.

But I'd prefer not to add `attach` or `+=`. I think that muddles the job of `Lifetime`—which should be just to notify observers when the lifetime has ended. I don't think we should give `Lifetime` any knowledge of `Disposable`.

**Comment by andersio** [comment]
I don't quite see the problem here. We can just treat a `Disposable` as a termination observer.

**Comment by mdiep** [comment]
Yeah, so I think writing `lifetime.observeEnded(disposable.dispose)` is fine. I don't think it requires extra sugar that couples the two concepts.

**Comment by andersio** [comment]
The deprecation of `ended` is reverted, while `observeEnded` is documented as the preferred API unless composition is needed.

File: Tests/ReactiveSwiftTests/LifetimeSpec.swift [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import Quick
 import Nimble
-import ReactiveSwift
+@testable import ReactiveSwift
```
Comment (sharplet) [inline Tests/ReactiveSwiftTests/LifetimeSpec.swift]: Does this still need to be `@testable`?
